File: src/lib/exerciseAreaGroups.js
/** Display order for routine / workout exercise sections (push-day style). */
export const AREA_GROUP_ORDER = ["chest", "back", "shoulders", "arms", "legs", "core", "other"];

export const AREA_GROUP_LABELS = {
  chest: "Chest",
  back: "Back",
  shoulders: "Shoulders",
  arms: "Arms & triceps",
  legs: "Legs",
  core: "Core",
  other: "Other",
};

export function normalizeAreaCategory(category) {
  const c = String(category || "other")
    .trim()
    .toLowerCase();
  if (AREA_GROUP_ORDER.includes(c)) return c;
  if (c.includes("triceps") || c.includes("biceps") || c === "arm") return "arms";
  if (c.includes("shoulder")) return "shoulders";
  if (c.includes("leg") || c.includes("quad") || c.includes("hamstring") || c.includes("calf"))
    return "legs";
  if (c.includes("chest") || c.includes("pec")) return "chest";
  if (c.includes("back") || c.includes("lat")) return "back";
  if (c.includes("core") || c.includes("abs") || c.includes("ab")) return "core";
  return "other";
}

/**
 * @template T
 * @param {T[]} exercises — items with optional `category`
 * @param {(item: T) => string} [getCategory]
 * @returns {{ area: string, label: string, exercises: T[] }[]}
 */
export function groupExercisesByArea(exercises, getCategory = ex => ex?.category) {
  const buckets = new Map();
  for (const ex of exercises || []) {
    const area = normalizeAreaCategory(getCategory(ex));
    if (!buckets.has(area)) buckets.set(area, []);
    buckets.get(area).push(ex);
  }

  const out = [];
  for (const area of AREA_GROUP_ORDER) {
    const items = buckets.get(area);
    if (items?.length) {
      out.push({ area, label: AREA_GROUP_LABELS[area], exercises: items });
      buckets.delete(area);
    }
  }
  for (const [area, items] of buckets) {
    if (items?.length) {
      out.push({
        area,
        label: AREA_GROUP_LABELS[area] || area,
        exercises: items,
      });
    }
  }
  return out;
}

/** Replace one area's exercises after within-group reorder; preserve other areas' order. */
export function mergeAreaReorder(fullList, area, reorderedInArea, getCategory = ex => ex?.category) {
  const groups = groupExercisesByArea(fullList, getCategory);
  const next = [];
  for (const g of groups) {
    next.push(...(g.area === area ? reorderedInArea : g.exercises));
  }
  return next;
}

File: ios/Logbook/Utilities/ExerciseAreaGroups.swift
import Foundation

enum ExerciseAreaGroups {
    static let order = ["chest", "back", "shoulders", "arms", "legs", "core", "other"]

    static let labels: [String: String] = [
        "chest": "Chest",
        "back": "Back",
        "shoulders": "Shoulders",
        "arms": "Arms & triceps",
        "legs": "Legs",
        "core": "Core",
        "other": "Other"
    ]

    static func normalizeAreaCategory(_ category: String?) -> String {
        let c = (category ?? "other").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if order.contains(c) { return c }
        if c.contains("triceps") || c.contains("biceps") || c == "arm" { return "arms" }
        if c.contains("shoulder") { return "shoulders" }
        if c.contains("leg") || c.contains("quad") || c.contains("hamstring") || c.contains("calf") { return "legs" }
        if c.contains("chest") || c.contains("pec") { return "chest" }
        if c.contains("back") || c.contains("lat") { return "back" }
        if c.contains("core") || c.contains("abs") || c.contains("ab") { return "core" }
        return "other"
    }

    struct AreaGroup: Identifiable {
        let area: String
        let label: String
        let exercises: [NativeExercise]
        var id: String { area }
    }

    static func groupExercises(_ exercises: [NativeExercise]) -> [AreaGroup] {
        var buckets: [String: [NativeExercise]] = [:]
        for exercise in exercises {
            let area = normalizeAreaCategory(exercise.category)
            buckets[area, default: []].append(exercise)
        }
        var groups: [AreaGroup] = []
        for area in order {
            guard let items = buckets[area], !items.isEmpty else { continue }
            groups.append(AreaGroup(area: area, label: labels[area] ?? area, exercises: items))
            buckets.removeValue(forKey: area)
        }
        for (area, items) in buckets where !items.isEmpty {
            groups.append(AreaGroup(area: area, label: labels[area] ?? area.capitalized, exercises: items))
        }
        return groups
    }
}
